### PR TITLE
HIP ALU ops overload cleanup try 2

### PR DESCRIPTION
**Pull request by Qazalin**
time to get rid of the overloads I originally had to put in here.

### CODE REVIEW

File: test/test_renderer.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    assert lang.code_for_op[UnaryOps.SIN]("a", dtypes.half) == "hsin((half)a)"
+    assert lang.code_for_op[UnaryOps.SIN]("a", dtypes.float) == "sin(a)"
+    assert lang.code_for_op[UnaryOps.SIN]("a", dtypes.float.vec(4)) == "float4(sin(a.x), sin(a.y), sin(a.z), sin(a.w))"
+
+    assert lang.code_for_op[UnaryOps.NEG]("a", dtypes.half) == "(-(half)a)"
+    assert lang.code_for_op[UnaryOps.NEG]("a", dtypes.bool) == "(!a)"
+    assert lang.code_for_op[UnaryOps.NEG]("a", dtypes.float) == "(-a)"
+
+    assert lang.code_for_op[BinaryOps.ADD]("a", "b", dtypes.float) == "(a+b)"
+    assert lang.code_for_op[BinaryOps.ADD]("a", "b", dtypes.half) == "__hadd((half)a,(half)b)"
```
Comment (geohot) [inline test/test_renderer.py:22]: So a test like this is very fragile. I would add this to uops instead and test for behavior

File: tinygrad/renderer/cstyle.py [hunk 2]
```diff
@@ -300,32 +297,18 @@ class HIPLanguage(CStyleLanguage):
...
+    UnaryOps.NEG: lambda a,dtype: f"(-(half){a})" if dtype == dtypes.half else CStyleLanguage().code_for_op[UnaryOps.NEG](a,dtype),
+    **{op: lambda a, b, dtype, op=op, fxn=fxn: f"__h{fxn}((half){a},(half){b})" if dtype == dtypes.half else CStyleLanguage().code_for_op[op](a,b,dtype) for op, fxn in zip([BinaryOps.ADD, BinaryOps.SUB, BinaryOps.MUL, BinaryOps.DIV, BinaryOps.CMPLT], ["add","sub","mul","div", "lt"])},
+    BinaryOps.MAX: lambda a,b,dtype: f"__hgt((half){a},(half){b})?(half){a}:(half){b}" if dtype == dtypes.half else f"float4(max({a}.x,{b}.x),max({a}.y,{b}.y),max({a}.z,{b}.z),max({a}.w,{b}.w))" if dtype == dtypes.float.vec(4) else f"max({a},{b})",
+    BinaryOps.MOD: lambda a,b,dtype: f"__hsub({a}, __hmul({b}, __float2half(floorf(__half2float({a}) / __half2float({b})))))" if dtype == dtypes.half else f"{a}%{b}",
```
Comment (geohot) [inline tinygrad/renderer/cstyle.py:309]: I think MOD should only be supported on ints

File: tinygrad/renderer/cstyle.py [hunk 3]
```diff
@@ -300,32 +297,18 @@ class HIPLanguage(CStyleLanguage):
...
+    TernaryOps.WHERE: lambda a,b,c,dtype: f"((float){a}!=0?{b}:{c})" if dtype != dtypes.half else f"(half)((float){a}!=0?{b}:{c})"
```
Comment (geohot) [inline tinygrad/renderer/cstyle.py:310]: This is hacks, see what @chenyuxyz is working on, a should always be a bool

**Comment by geohot** [comment]
If you want to check HIP you can't run in CI, maybe look at the RDNA3 asm?

**Comment by Qazalin** [comment]
yeah right I'm just compiling it with docker and generating a diff between these kernels and master

File: tinygrad/renderer/cstyle.py [hunk 4]
```diff
@@ -297,17 +294,16 @@ class HIPLanguage(CStyleLanguage):
...
+  code_for_op = {
+    **CStyleLanguage().code_for_op, **{op: lambda a,dtype,op=op: f"h{CStyleLanguage().code_for_op[op]('(half)'+a,dtype)}" if dtype == dtypes.half else f"float4({CStyleLanguage().code_for_op[op](a+'.x',dtype)}, {CStyleLanguage().code_for_op[op](a+'.y',dtype)}, {CStyleLanguage().code_for_op[op](a+'.z',dtype)}, {CStyleLanguage().code_for_op[op](a+'.w',dtype)})" if dtype == dtypes.float.vec(4) else f"{CStyleLanguage().code_for_op[op](a,dtype)}" for op in UnaryOps}, # noqa: E501
+    UnaryOps.NEG: lambda a,dtype: f"(-(half){a})" if dtype == dtypes.half else CStyleLanguage().code_for_op[UnaryOps.NEG](a,dtype), # noqa: E501
+    **{op: lambda a, b, dtype, op=op, fxn=fxn: f"__h{fxn}((half){a},(half){b})" if dtype == dtypes.half else CStyleLanguage().code_for_op[op](a,b,dtype) for op, fxn in zip([BinaryOps.ADD, BinaryOps.SUB, BinaryOps.MUL, BinaryOps.DIV], ["add","sub","mul","div"])}, # noqa: E501
+    BinaryOps.MAX: lambda a,b,dtype: f"__hgt((half){a},(half){b})?(half){a}:(half){b}" if dtype == dtypes.half else f"float4(max({a}.x,{b}.x),max({a}.y,{b}.y),max({a}.z,{b}.z),max({a}.w,{b}.w))" if dtype == dtypes.float.vec(4) else f"max({a},{b})",# noqa: E501
```
Comment (geohot) [inline tinygrad/renderer/cstyle.py:302]: Careful, this runs things twice with __hgt.

And why do you need float4 here? max shouldn't be done on float4

**Comment by geohot** [comment]
Btw, those lines are completely unreadable and violate the 150 character limit. No new code can be merged with > 150 char lines.

**Pull request closed without merging**